File: JDK11/opentraceability/src/main/java/opentraceability/models/masterdata/kdes/masterdatakdestring.java
package opentraceability.models.masterdata.kdes;

import opentraceability.interfaces.IMasterDataKDE;
import org.json.JSONObject;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import javax.xml.parsers.DocumentBuilderFactory;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;

public class MasterDataKDEString extends MasterDataKDEBase implements IMasterDataKDE {

    public Type valueType = String.class;

    public String value = null;
    public String type = null;
    public Map<String, String> attributes = new HashMap<>();

    public MasterDataKDEString() {
    }

    public MasterDataKDEString(Type valueType) {
        // Default constructor
        this.valueType = valueType;
    }

    public MasterDataKDEString(String ns, String name) {
        super.namespace = ns;
        super.name = name;
    }

    public JSONObject getJson() {
        if (value == null || value.isEmpty()) {
            return null;
        } else {
            return new JSONObject(value);
        }
    }


    public Element getXml() throws Exception {
        if (value == null || value.isEmpty()) {
            return null;
        } else {
            String xname = namespace + name;
            Document document = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
            Element element = document.createElement(xname);
            element.setTextContent(value);

            // Set the xsi type...
            for (Map.Entry<String, String> entry : attributes.entrySet()) {
                String key = entry.getKey();
                String val = entry.getValue();
                element.setAttribute(key, val);
            }

            return element;
        }
    }


    public void setFromJson(JSONObject json) {
        value = json.toString();
    }

    public void setFromXml(Element xml) {
        value = xml.getTextContent();

        for (int i = 0; i < xml.getAttributes().getLength(); i++) {
            String key = xml.getAttributes().item(i).getNodeName();
            String val = xml.getAttributes().item(i).getNodeValue();
            attributes.put(key, val);
        }
    }

    @Override
    public String toString() {
        return value == null ? "" : value;
    }

    @Override
    public void setFromGS1WebVocabJson(JSONObject json) {
        value = json.toString();
    }

    public JSONObject getGS1WebVocabJson() {
        if (value != null) {
            return new JSONObject(value);
        } else {
            return null;
        }
    }

    @Override
    public void setFromEPCISXml(Element xml) {
        name = xml.getAttribute("id") != null ? xml.getAttribute("id") : "";
        value = xml.getTextContent();
    }

    @Override
    public Element getEPCISXml() throws Exception {
        if (value != null) {
            Document document = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
            Element element = document.createElement("attribute");
            element.setAttribute("id", name);
            element.setTextContent(value);
            return element;
        } else {
            return null;
        }
    }
}

File: CSharp/OpenTraceability/Models/MasterData/KDEs/MasterDataKDEString.cs
using Newtonsoft.Json.Linq;
using OpenTraceability.Interfaces;
using OpenTraceability.Models.MasterData.KDEs;
using System.Xml.Linq;

namespace OpenTraceability.Models.Events.KDEs
{
    public class MasterDataKDEString : MasterDataKDEBase, IMasterDataKDE
    {
        public Type ValueType => typeof(string);
        public string Value { get; set; }
        public string Type { get; set; }
        Dictionary<string, string> Attributes { get; set; } = new Dictionary<string, string>();

        internal MasterDataKDEString()
        {

        }

        public MasterDataKDEString(string ns, string name)
        {
            this.Namespace = ns;
            this.Name = name;
        }

        public JToken GetJson()
        {
            if (string.IsNullOrWhiteSpace(Value))
            {
                return null;
            }
            else
            {
                return JToken.FromObject(Value);
            }
        }

        public XElement GetXml()
        {
            if (string.IsNullOrWhiteSpace(Value))
            {
                return null;
            }
            else
            {
                XName xname = (XNamespace)Namespace + Name;
                XElement x = new XElement(xname, Value);

                // set the xsi type...
                foreach (var kvp in Attributes)
                {
                    x.Add(new XAttribute(kvp.Key, kvp.Value));
                }

                return x;
            }
        }

        public void SetFromJson(JToken json)
        {
            this.Value = json.ToString();
        }

        public void SetFromXml(XElement xml)
        {
            this.Value = xml.Value;

            foreach (XAttribute xatt in xml.Attributes())
            {
                Attributes.Add(xatt.Name.ToString(), xatt.Value);
            }
        }

        public override string ToString()
        {
            return Value ?? string.Empty;
        }

        public void SetFromGS1WebVocabJson(JToken json)
        {
            this.Value = json.ToString();
        }

        public JToken GetGS1WebVocabJson()
        {
            if (this.Value != null)
            {
                return JToken.FromObject(this.Value);
            }
            else
            {
                return null;
            }
        }

        public void SetFromEPCISXml(XElement xml)
        {
            Name = xml.Attribute("id")?.Value ?? string.Empty;
            Value = xml.Value;
        }

        public XElement GetEPCISXml()
        {
            if (Value != null)
            {
                XElement x = new XElement("attribute");
                x.Add(new XAttribute("id", Name));
                x.Value = Value;
                return x;
            }
            else
            {
                return null;
            }
        }
    }
}